README.md
<a href="https://codecov.io/github/jalvaradosegura/date-operations" >
    <img src="https://codecov.io/github/jalvaradosegura/date-operations/branch/main/graph/badge.svg?token=TO89NFDQ79"/>
</a>

---

# date-operations
🗓 Easy to use functions for common date operations

---

## Installation

```console
pip install date-operations
```

## Usage
### Days between 2 dates
Get the amount of days between 2 dates:

```py
from date_operations import days_between

days_between("2023-01-01", "2023-01-02")  # 1
days_between("01 01 2023", "01/01/2024")  # 365

# You can omit the day:
days_between("jan-2023", "01/02/2023")  # 31
```

### Months between 2 dates
Get the amount of full months between 2 dates:

```py
from date_operations import months_between

months_between("01 05 2023", "31 05 2023")  # 0
months_between("01-05-2023", "01-06-2023")  # 1
months_between("10/05/2022", "10/05/2023")  # 12
months_between("10-may-2022", "09-may-2023")  # 11

# You can omit the day:
months_between("05-2023", "06-2023")  # 1
```

### Months started between 2 dates
Get the amount of months started between 2 dates:

```py
from date_operations import months_started_between

months_started_between("31 05 2023", "01 06 2023")  # 1
months_started_between("01-05-2023", "01-06-2023")  # 1
months_started_between("31/12/2022", "01/01/2023")  # 1
```

### Years between 2 dates
Get the amount of full years between 2 dates:

```py
from date_operations import years_between

years_between("15 01 2022", "15 01 2023")  # 1
years_between("15-01-2022", "14-01-2023")  # 0
years_between("31/12/2022", "01/01/2023")  # 0

# You can omit the day:
years_between("05/2023", "06/2023")  # 0
years_between("05/2023", "01/2024")  # 0
years_between("05/2023", "05/2024")  # 1
```

### Years started between 2 dates
Get the amount of years started between 2 dates:

```py
from date_operations import years_started_between

years_started_between("31 12 2022", "01 01 2023")  # 1
years_started_between("01-01-2023", "31-12-2023")  # 0
years_started_between("31/12/2022", "01/01/2024")  # 2
```

### Date formats
Date formats are inferred given [this priority list](https://github.com/jalvaradosegura/date-operations/blob/main/date_operations/settings.py). If the format of your date is not within the list you can add it or explicitly indicate the format of a given date. The formats added to the list are given priority.

```py
from date_operations import months_between

months_between("01-01", "01-02", format_1="%d-%m", format_2="%d-%m")  # 1
months_between("01-01", "01-02", extra_formats=["%d-%m"])  # 1
```
> The same parameters are used for all the functions.


release-notes.md
# Release Notes

## Latest Changes

* ⬆ [pre-commit.ci] pre-commit autoupdate. PR [#21](https://github.com/jalvaradosegura/date-operations/pull/21) by [@pre-commit-ci[bot]](https://github.com/apps/pre-commit-ci).
* ⬆ [pre-commit.ci] pre-commit autoupdate. PR [#20](https://github.com/jalvaradosegura/date-operations/pull/20) by [@pre-commit-ci[bot]](https://github.com/apps/pre-commit-ci).
* 📝 Add more library details to `pyproject.toml`. PR [#19](https://github.com/jalvaradosegura/date-operations/pull/19) by [@jalvaradosegura](https://github.com/jalvaradosegura).
* 📝 Add installation section to `README.md`. PR [#18](https://github.com/jalvaradosegura/date-operations/pull/18) by [@jalvaradosegura](https://github.com/jalvaradosegura).

## 0.2.0

### Features
* ✨ Turn decorator into a simple function. PR [#12](https://github.com/jalvaradosegura/date-operations/pull/12) by [@jalvaradosegura](https://github.com/jalvaradosegura).
* ✨ Add a decorator to guess the date formats. PR [#10](https://github.com/jalvaradosegura/date-operations/pull/10) by [@jalvaradosegura](https://github.com/jalvaradosegura).
* ✨ Add function to guess format of given date [private function]. PR [#9](https://github.com/jalvaradosegura/date-operations/pull/9) by [@jalvaradosegura](https://github.com/jalvaradosegura).

### Docs
* 📝 Add `__version__` variable to the library. PR [#17](https://github.com/jalvaradosegura/date-operations/pull/17) by [@jalvaradosegura](https://github.com/jalvaradosegura).
* 📝 Add details to the usage section of `README.md`. PR [#16](https://github.com/jalvaradosegura/date-operations/pull/16) by [@jalvaradosegura](https://github.com/jalvaradosegura).
* 📝 Add `LICENSE`. PR [#14](https://github.com/jalvaradosegura/date-operations/pull/14) by [@jalvaradosegura](https://github.com/jalvaradosegura).
* 📝 Add docstrings to main functions. PR [#13](https://github.com/jalvaradosegura/date-operations/pull/13) by [@jalvaradosegura](https://github.com/jalvaradosegura).

### Fixes
* 🐛 Add `extra_formats` to the default list. PR [#15](https://github.com/jalvaradosegura/date-operations/pull/15) by [@jalvaradosegura](https://github.com/jalvaradosegura).
* 🐛 Fix `UnboundLocalError` in the decorator. PR [#11](https://github.com/jalvaradosegura/date-operations/pull/11) by [@jalvaradosegura](https://github.com/jalvaradosegura).

## 0.1.0

### Features
* ✨ Add "started" functions. PR [#3](https://github.com/jalvaradosegura/date-operations/pull/3) by [@jalvaradosegura](https://github.com/jalvaradosegura).
* ✨ Add functions to calculate differences. PR [#2](https://github.com/jalvaradosegura/date-operations/pull/2) by [@jalvaradosegura](https://github.com/jalvaradosegura).

### Docs
* 📝 Add basic usage of the library. PR [#5](https://github.com/jalvaradosegura/date-operations/pull/5) by [@jalvaradosegura](https://github.com/jalvaradosegura).

### Fixes
* 🎨 Change `pre-commit` autofix message. PR [#8](https://github.com/jalvaradosegura/date-operations/pull/8) by [@jalvaradosegura](https://github.com/jalvaradosegura).
* 🎨 Add `pre-commit`. PR [#7](https://github.com/jalvaradosegura/date-operations/pull/7) by [@jalvaradosegura](https://github.com/jalvaradosegura).
* 🎨 Allow direct import of the main functions. PR [#4](https://github.com/jalvaradosegura/date-operations/pull/4) by [@jalvaradosegura](https://github.com/jalvaradosegura).

### Internal
* 💚 Add coverage action. PR [#6](https://github.com/jalvaradosegura/date-operations/pull/6) by [@jalvaradosegura](https://github.com/jalvaradosegura).


tests/test_main.py
from typing import List, Optional, Tuple

import pytest

from date_operations import (
    days_between,
    months_between,
    months_started_between,
    years_between,
    years_started_between,
)
from date_operations.main import _guess_date_format


@pytest.mark.parametrize(
    "dates, result",
    [
        [("2023-01-01", "2023-01-11"), 10],
        [("2023-01-01", "2023-01-01"), 0],
        [("2023-01-01", "2023-02-01"), 31],
        [("2023-01-01", "2024-01-01"), 365],
    ],
)
def test_days_between(dates: Tuple[str, str], result: int):
    date_1, date_2 = dates
    assert days_between(date_1, date_2) == result


@pytest.mark.parametrize(
    "dates, result",
    [
        [("2023 01 01", "20230111"), 10],
        [("2023/01/01", "01-01-2023"), 0],
        [("01/01/2023", "01-feb-2023"), 31],
        [("01 jan 2023", "01/jan/2024"), 365],
    ],
)
def test_days_between_guess_formats(dates: Tuple[str, str], result: int):
    date_1, date_2 = dates
    assert days_between(date_1, date_2) == result


@pytest.mark.parametrize(
    "dates, result",
    [
        [("2023-01-01", "2023-02-01"), 1],
        [("2023-02-01", "2023-01-01"), 1],
        [("2023-01-01", "2023-01-01"), 0],
        [("2023-01-01", "2024-01-01"), 12],
        [("2023-01-31", "2023-02-01"), 0],
    ],
)
def test_months_between(dates: Tuple[str, str], result: int):
    date_1, date_2 = dates
    assert months_between(date_1, date_2) == result


@pytest.mark.parametrize(
    "dates, result",
    [
        [("2023 01 01", "20230201"), 1],
        [("2023/02/01", "01-01-2023"), 1],
        [("2023-01", "2023-01"), 0],
        [("01-2023", "jan-2024"), 12],
        [("jan/2023", "feb 2023"), 1],
    ],
)
def test_months_between_guess_format(dates: Tuple[str, str], result: int):
    date_1, date_2 = dates
    assert months_between(date_1, date_2) == result


@pytest.mark.parametrize(
    "dates, result",
    [
        [("2023-01-31", "2023-02-01"), 1],
        [("2023-02-01", "2023-01-01"), 1],
        [("2023-12-31", "2024-01-01"), 1],
        [("2024-01-01", "2023-12-31"), 1],
        [("2023-01-01", "2023-02-01"), 1],
        [("2023-02-01", "2023-01-01"), 1],
        [("2023-01-01", "2023-01-01"), 0],
        [("2023-01-01", "2024-01-01"), 12],
    ],
)
def test_months_started_between(dates: Tuple[str, str], result: int):
    date_1, date_2 = dates
    assert months_started_between(date_1, date_2) == result


@pytest.mark.parametrize(
    "dates, result",
    [
        [("2023-01-31", "2023/02/01"), 1],
        [("2023 02 01", "01-2023"), 1],
        [("20231231", "jan 2024"), 1],
    ],
)
def test_months_started_between_guess_format(dates: Tuple[str, str], result: int):
    date_1, date_2 = dates
    assert months_started_between(date_1, date_2) == result


@pytest.mark.parametrize(
    "dates, result",
    [
        [("2023-01-01", "2024-01-01"), 1],
        [("2023-01-01", "2022-01-01"), 1],
        [("2023-01-01", "2023-01-01"), 0],
        [("2023-01-01", "2030-01-01"), 7],
        [("2023-12-31", "2024-01-01"), 0],
    ],
)
def test_years_between(dates: Tuple[str, str], result: int):
    date_1, date_2 = dates
    assert years_between(date_1, date_2) == result


@pytest.mark.parametrize(
    "dates, result",
    [
        [("2023/01/01", "2024 01 01"), 1],
        [("20230101", "2022-01"), 1],
        [("2023-01", "202301"), 0],
        [("2023-01-01", "2030-01"), 7],
        [("2023-12", "2024-01"), 0],
        [("2023-12", "2024-12"), 1],
    ],
)
def test_years_between_guess_format(dates: Tuple[str, str], result: int):
    date_1, date_2 = dates
    assert years_between(date_1, date_2) == result


@pytest.mark.parametrize(
    "dates, result",
    [
        [("2023-12-31", "2024-01-01"), 1],
        [("2023-01-01", "2022-12-31"), 1],
        [("2023-01-01", "2030-01-01"), 7],
        [("2023-01-01", "2024-01-01"), 1],
        [("2023-01-01", "2022-01-01"), 1],
        [("2023-01-01", "2023-01-01"), 0],
        [("2023-01-01", "2030-01-01"), 7],
    ],
)
def test_years_started_between(dates: Tuple[str, str], result: int):
    date_1, date_2 = dates
    assert years_started_between(date_1, date_2) == result


@pytest.mark.parametrize(
    "dates, result",
    [
        [("2023-12", "2024-01"), 1],
        [("2023 01 01", "2022/12/31"), 1],
        [("jan-2023", "01-01-2030"), 7],
    ],
)
def test_years_started_between_guess_format(dates: Tuple[str, str], result: int):
    date_1, date_2 = dates
    assert years_started_between(date_1, date_2) == result


@pytest.mark.parametrize(
    "date, date_format",
    [
        ["10-10-2010", "%d-%m-%Y"],
        ["10/10/2010", "%d/%m/%Y"],
        ["10 10 2010", "%d %m %Y"],
        ["10102010", "%d%m%Y"],
        ["2010-10-10", "%Y-%m-%d"],
        ["2010/10/10", "%Y/%m/%d"],
        ["10-2010", "%m-%Y"],
        ["10/2010", "%m/%Y"],
        ["10 2010", "%m %Y"],
        ["102010", "%m%Y"],
        ["2010-10", "%Y-%m"],
        ["2010/10", "%Y/%m"],
        ["2010 10", "%Y %m"],
        ["201010", "%Y%m"],
        ["10-oct-2010", "%d-%b-%Y"],
        ["10/oct/2010", "%d/%b/%Y"],
        ["10 oct 2010", "%d %b %Y"],
        ["10oct2010", "%d%b%Y"],
        ["oct-2010", "%b-%Y"],
        ["oct/2010", "%b/%Y"],
        ["oct 2010", "%b %Y"],
        ["oct2010", "%b%Y"],
        ["unknown", None],
    ],
)
def test_guess_date_format(date: str, date_format: Optional[str]):
    assert _guess_date_format(date) == date_format


@pytest.mark.parametrize(
    "date, date_format, extra_formats",
    [
        ["2010 10 10", "%Y %m %d", ["%Y %m %d"]],
        ["20101010", "%Y%m%d", ["%Y%m%d"]],
        ["10 10 2010", "%m %d %Y", ["%m %d %Y"]],
        ["10102010", "%m%d%Y", ["%m%d%Y"]],
        ["10-10-2010", "%m-%d-%Y", ["%m-%d-%Y"]],
        ["10/10/2010", "%m/%d/%Y", ["%m/%d/%Y"]],
    ],
)
def test_guess_date_format_with_extra_formats(
    date: str, date_format: str, extra_formats: List[str]
):
    assert _guess_date_format(date, extra_formats) == date_format


def test_no_guess_date_1():
    date_1 = "invalid-date"
    date_2 = "2023-01-11"

    with pytest.raises(
        ValueError,
        match=f"Couldn't guess the date format for date: {date_1}.",
    ):
        days_between(date_1, date_2)


def test_no_guess_date_2():
    date_1 = "2023-01-11"
    date_2 = "invalid-date"

    with pytest.raises(
        ValueError,
        match=f"Couldn't guess the date format for date: {date_2}.",
    ):
        days_between(date_1, date_2)


def test_use_custom_formats():
    assert days_between("01", "09", format_1="%d", format_2="%d") == 8
    assert months_between("01-01", "01-02", format_1="%d-%m", format_2="%d-%m") == 1
    assert (
        months_started_between("01/01", "01/02", format_1="%d/%m", format_2="%d/%m")
        == 1
    )
    assert years_between("2000", "2023-31", format_1="%Y", format_2="%Y-%d") == 23
    assert (
        years_started_between("2000", "2023 31", format_1="%Y", format_2="%Y %d") == 23
    )


def test_use_extra_formats():
    """
    If you are aware of possible date formats that are not included in the default
    format list, you can add them using the `extra_formats` parameter. These additional
    formats will be given priority in the date format detection process.
    """
    assert days_between("01", "09", extra_formats=["%d"]) == 8
    assert months_between("01-01", "01-02", extra_formats=["%d-%m"]) == 1
    assert months_started_between("01/01", "01/02", extra_formats=["%d/%m"]) == 1
    assert years_between("2000", "2023-31", extra_formats=["%Y", "%Y-%d"]) == 23
    assert years_started_between("2000", "2023 31", extra_formats=["%Y", "%Y %d"]) == 23


date_operations/main.py
from datetime import datetime
from typing import List, Optional

from dateutil.relativedelta import relativedelta

from .settings import date_format_list


def days_between(
    date_1: str,
    date_2: str,
    format_1: Optional[str] = None,
    format_2: Optional[str] = None,
    extra_formats: Optional[List[str]] = None,
) -> int:
    """
    Returns the number of days between two dates.

    This function tries to guess the format of the dates automatically based on a
    list of formats. You can also explicitly provide the format for each date.

    Moreover, you can indicate extra formats that are added to the list of formats.
    These extra formats are given more priority than the default ones.

    In addition, the function considers a special case where the input dates are missing
    the "day" but have the "year" and the "month". In this case, the function considers
    the dates as the first day of the month.
    """
    return abs(
        _to_datetime(date_1, format_1, extra_formats)
        - _to_datetime(date_2, format_2, extra_formats)
    ).days


def months_between(
    date_1: str,
    date_2: str,
    format_1: Optional[str] = None,
    format_2: Optional[str] = None,
    extra_formats: Optional[List[str]] = None,
) -> int:
    """
    Returns the number of full months between two dates.

    Full months in this case means:
    - months_between("01 05 2023", "31 05 2023")  # 0
    - months_between("01-05-2023", "01-06-2023")  # 1
    - months_between("10/05/2022", "10/05/2023")  # 12
    - months_between("10-may-2022", "09-may-2023")  # 11

    You can omit the day:
    - months_between("05-2023", "06-2023")  # 1

    This function tries to guess the format of the dates automatically based on a
    list of formats. You can also explicitly provide the format for each date.

    Moreover, you can indicate extra formats that are added to the list of formats.
    These extra formats are given more priority than the default ones.
    """
    delta = relativedelta(
        _to_datetime(date_1, format_1, extra_formats),
        _to_datetime(date_2, format_2, extra_formats),
    )
    return abs(delta.months + (12 * delta.years))


def months_started_between(
    date_1: str,
    date_2: str,
    format_1: Optional[str] = None,
    format_2: Optional[str] = None,
    extra_formats: Optional[List[str]] = None,
) -> int:
    """
    Returns the number of months started between two dates.

    Months started in this case means:
    - months_started_between("31-05-2023", "01-06-2023")  # 1
    - months_started_between("01-05-2023", "01-06-2023")  # 1
    - months_started_between("31-12-2022", "01-01-2023")  # 1

    You can omit the day and in that case this function behaves just like
    `months_between`.

    This function tries to guess the format of the dates automatically based on a
    list of formats. You can also explicitly provide the format for each date.

    Moreover, you can indicate extra formats that are added to the list of formats.
    These extra formats are given more priority than the default ones.
    """
    ddate_1 = _to_datetime(date_1, format_1, extra_formats)
    ddate_2 = _to_datetime(date_2, format_2, extra_formats)
    months_started = ddate_1.month - ddate_2.month
    years_started = ddate_1.year - ddate_2.year
    return abs(years_started * 12 + months_started)


def years_between(
    date_1: str,
    date_2: str,
    format_1: Optional[str] = None,
    format_2: Optional[str] = None,
    extra_formats: Optional[List[str]] = None,
) -> int:
    """
    Returns the number of full years between two dates.

    Full years in this case means:
    - years_between("15 01 2022", "15 01 2023")  # 1
    - years_between("15-01-2022", "14-01-2023")  # 0
    - years_between("31/12/2022", "01/01/2023")  # 0

    You can omit the day:
    - years_between("05/2023", "06/2023")  # 0
    - years_between("05/2023", "01/2024")  # 0
    - years_between("05/2023", "05/2024")  # 1

    This function tries to guess the format of the dates automatically based on a
    list of formats. You can also explicitly provide the format for each date.

    Moreover, you can indicate extra formats that are added to the list of formats.
    These extra formats are given more priority than the default ones.
    """
    delta = relativedelta(
        _to_datetime(date_1, format_1, extra_formats),
        _to_datetime(date_2, format_2, extra_formats),
    )
    return abs(delta.years)


def years_started_between(
    date_1: str,
    date_2: str,
    format_1: Optional[str] = None,
    format_2: Optional[str] = None,
    extra_formats: Optional[List[str]] = None,
) -> int:
    """
    Returns the number of years started between two dates.

    Years started in this case means:
    - years_started_between("31-12-2022", "01-01-2023")  # 1
    - years_started_between("01-01-2023", "31-12-2023")  # 0
    - years_started_between("31-12-2022", "01-01-2024")  # 2

    You can omit the day:
    - years_started_between("12-2022", "01-2024")  # 2

    This function tries to guess the format of the dates automatically based on a
    list of formats. You can also explicitly provide the format for each date.

    Moreover, you can indicate extra formats that are added to the list of formats.
    These extra formats are given more priority than the default ones.
    """
    years_started = (
        _to_datetime(date_1, format_1, extra_formats).year
        - _to_datetime(date_2, format_2, extra_formats).year
    )
    return abs(years_started)


def _to_datetime(
    date: str, _format: Optional[str], extra_formats: Optional[List[str]]
) -> datetime:
    if _format is None:
        _format = _guess_date_format(date, extra_formats)

    if _format is None:
        raise ValueError(f"Couldn't guess the date format for date: {date}.")

    return datetime.strptime(date, _format)


def _guess_date_format(
    date: str, extra_formats: Optional[List[str]] = None
) -> Optional[str]:
    if extra_formats is None:
        extra_formats = []

    date_formats = extra_formats + date_format_list

    for date_format in date_formats:
        try:
            datetime.strptime(date, date_format)
        except ValueError:
            pass
        else:
            return date_format

    return None
